### PR TITLE
Base64 decoder

### PR DESCRIPTION
**Pull request by asgibbons**
Adding x86-64 code for base64 decoding.  The algorithm uses AVX512 VBMI instructions.

Also fixed what I believe is a bug in Base64.java.  The decoded length calculation should be ceil(dl/3)*4.

Benchmark results:
Adding this intrinsic code has a negative effect on performance for buffers shorter than ~112 bytes (up to ~2x worse).  However for buffer sizes > ~112, performance increases by 1.2x to just under 15x.
MIME performance is consistently worse (again, up to 2x worse).
Error detection is essentially unchanged.

**Without this change**:
```
Benchmark                                                   (errorIndex)  (lineSize)  (maxNumBytes)  Mode  Samples       Score  Score error  Units
o.o.m.b.j.u.Base64Decode.testBase64Decode                            144           4              1  avgt      200      15.357        0.035  ns/op
o.o.m.b.j.u.Base64Decode.testBase64Decode                            144           4              3  avgt      200      17.002        0.214  ns/op
o.o.m.b.j.u.Base64Decode.testBase64Decode                            144           4              7  avgt      200      20.597        0.089  ns/op
o.o.m.b.j.u.Base64Decode.testBase64Decode                            144           4             32  avgt      200      34.213        0.225  ns/op
o.o.m.b.j.u.Base64Decode.testBase64Decode                            144           4             64  avgt      200      54.432        0.183  ns/op
o.o.m.b.j.u.Base64Decode.testBase64Decode                            144           4             80  avgt      200      66.395        0.350  ns/op
o.o.m.b.j.u.Base64Decode.testBase64Decode                            144           4             96  avgt      200      73.162        0.314  ns/op
o.o.m.b.j.u.Base64Decode.testBase64Decode                            144           4            112  avgt      200      84.927        0.287  ns/op
o.o.m.b.j.u.Base64Decode.testBase64Decode                            144           4            512  avgt      200     288.810        2.893  ns/op
o.o.m.b.j.u.Base64Decode.testBase64Decode                            144           4           1000  avgt      200     560.481        1.439  ns/op
o.o.m.b.j.u.Base64Decode.testBase64Decode                            144           4          20000  avgt      200    9530.278      107.781  ns/op
o.o.m.b.j.u.Base64Decode.testBase64Decode                            144           4          50000  avgt      200   24552.237      298.532  ns/op
o.o.m.b.j.u.Base64Decode.testBase64MIMEDecode                        144           4              1  avgt      200      22.865        0.522  ns/op
o.o.m.b.j.u.Base64Decode.testBase64MIMEDecode                        144           4              3  avgt      200      27.793        0.710  ns/op
o.o.m.b.j.u.Base64Decode.testBase64MIMEDecode                        144           4              7  avgt      200      44.741        1.227  ns/op
o.o.m.b.j.u.Base64Decode.testBase64MIMEDecode                        144           4             32  avgt      200     142.688        3.730  ns/op
o.o.m.b.j.u.Base64Decode.testBase64MIMEDecode                        144           4             64  avgt      200     256.899        8.234  ns/op
o.o.m.b.j.u.Base64Decode.testBase64MIMEDecode                        144           4             80  avgt      200     311.599       10.242  ns/op
o.o.m.b.j.u.Base64Decode.testBase64MIMEDecode                        144           4             96  avgt      200     364.001       12.745  ns/op
o.o.m.b.j.u.Base64Decode.testBase64MIMEDecode                        144           4            112  avgt      200     472.883       13.226  ns/op
o.o.m.b.j.u.Base64Decode.testBase64MIMEDecode                        144           4            512  avgt      200    1814.963       67.365  ns/op
o.o.m.b.j.u.Base64Decode.testBase64MIMEDecode                        144           4           1000  avgt      200    3623.500      128.688  ns/op
o.o.m.b.j.u.Base64Decode.testBase64MIMEDecode                        144           4          20000  avgt      200   70484.092     2391.142  ns/op
o.o.m.b.j.u.Base64Decode.testBase64MIMEDecode                        144           4          50000  avgt      200  191732.336     7280.298  ns/op
o.o.m.b.j.u.Base64Decode.testBase64WithErrorInputsDecode             144           4              1  avgt      200    1531.022       30.085  ns/op
o.o.m.b.j.u.Base64Decode.testBase64WithErrorInputsDecode             144           4              3  avgt      200    1306.593       41.400  ns/op
o.o.m.b.j.u.Base64Decode.testBase64WithErrorInputsDecode             144           4              7  avgt      200    1238.114       40.971  ns/op
o.o.m.b.j.u.Base64Decode.testBase64WithErrorInputsDecode             144           4             32  avgt      200    1346.459       49.189  ns/op
o.o.m.b.j.u.Base64Decode.testBase64WithErrorInputsDecode             144           4             64  avgt      200    1195.280       27.871  ns/op
o.o.m.b.j.u.Base64Decode.testBase64WithErrorInputsDecode             144           4             80  avgt      200    1469.004       39.480  ns/op
o.o.m.b.j.u.Base64Decode.testBase64WithErrorInputsDecode             144           4             96  avgt      200    1434.475       45.145  ns/op
o.o.m.b.j.u.Base64Decode.testBase64WithErrorInputsDecode             144           4            112  avgt      200    1440.061       45.011  ns/op
o.o.m.b.j.u.Base64Decode.testBase64WithErrorInputsDecode             144           4            512  avgt      200    1362.794       54.234  ns/op
o.o.m.b.j.u.Base64Decode.testBase64WithErrorInputsDecode             144           4           1000  avgt      200    1426.065       72.725  ns/op
o.o.m.b.j.u.Base64Decode.testBase64WithErrorInputsDecode             144           4          20000  avgt      200    1398.441       46.695  ns/op
o.o.m.b.j.u.Base64Decode.testBase64WithErrorInputsDecode             144           4          50000  avgt      200    1409.408       56.964  ns/op
```

**With this change:**
```
Benchmark                                                   (errorIndex)  (lineSize)  (maxNumBytes)  Mode  Samples       Score  Score error  Units
o.o.m.b.j.u.Base64Decode.testBase64Decode                            144           4              1  avgt      200      15.303        0.023  ns/op
o.o.m.b.j.u.Base64Decode.testBase64Decode                            144           4              3  avgt      200      16.629        0.192  ns/op
o.o.m.b.j.u.Base64Decode.testBase64Decode                            144           4              7  avgt      200      27.558        0.351  ns/op
o.o.m.b.j.u.Base64Decode.testBase64Decode                            144           4             32  avgt      200      91.265        0.743  ns/op
o.o.m.b.j.u.Base64Decode.testBase64Decode                            144           4             64  avgt      200     100.330        0.430  ns/op
o.o.m.b.j.u.Base64Decode.testBase64Decode                            144           4             80  avgt      200      84.671        0.491  ns/op
o.o.m.b.j.u.Base64Decode.testBase64Decode                            144           4             96  avgt      200      82.425        0.360  ns/op
o.o.m.b.j.u.Base64Decode.testBase64Decode                            144           4            112  avgt      200      70.568        0.603  ns/op
o.o.m.b.j.u.Base64Decode.testBase64Decode                            144           4            512  avgt      200      41.687        0.239  ns/op
o.o.m.b.j.u.Base64Decode.testBase64Decode                            144           4           1000  avgt      200      50.035        0.234  ns/op
o.o.m.b.j.u.Base64Decode.testBase64Decode                            144           4          20000  avgt      200     544.277        7.387  ns/op
o.o.m.b.j.u.Base64Decode.testBase64Decode                            144           4          50000  avgt      200    1649.279       37.060  ns/op
o.o.m.b.j.u.Base64Decode.testBase64MIMEDecode                        144           4              1  avgt      200      22.864        0.530  ns/op
o.o.m.b.j.u.Base64Decode.testBase64MIMEDecode                        144           4              3  avgt      200      36.221        0.833  ns/op
o.o.m.b.j.u.Base64Decode.testBase64MIMEDecode                        144           4              7  avgt      200      68.281        1.599  ns/op
o.o.m.b.j.u.Base64Decode.testBase64MIMEDecode                        144           4             32  avgt      200     209.307        7.347  ns/op
o.o.m.b.j.u.Base64Decode.testBase64MIMEDecode                        144           4             64  avgt      200     428.140       14.870  ns/op
o.o.m.b.j.u.Base64Decode.testBase64MIMEDecode                        144           4             80  avgt      200     582.169       16.083  ns/op
o.o.m.b.j.u.Base64Decode.testBase64MIMEDecode                        144           4             96  avgt      200     696.114       20.352  ns/op
o.o.m.b.j.u.Base64Decode.testBase64MIMEDecode                        144           4            112  avgt      200     668.192       22.979  ns/op
o.o.m.b.j.u.Base64Decode.testBase64MIMEDecode                        144           4            512  avgt      200    4032.949      133.964  ns/op
o.o.m.b.j.u.Base64Decode.testBase64MIMEDecode                        144           4           1000  avgt      200    8244.454      303.384  ns/op
o.o.m.b.j.u.Base64Decode.testBase64MIMEDecode                        144           4          20000  avgt      200  177529.136     7124.951  ns/op
o.o.m.b.j.u.Base64Decode.testBase64MIMEDecode                        144           4          50000  avgt      200  401161.776    19129.292  ns/op
o.o.m.b.j.u.Base64Decode.testBase64WithErrorInputsDecode             144           4              1  avgt      200    1255.647       47.682  ns/op
o.o.m.b.j.u.Base64Decode.testBase64WithErrorInputsDecode             144           4              3  avgt      200    1301.833       47.020  ns/op
o.o.m.b.j.u.Base64Decode.testBase64WithErrorInputsDecode             144           4              7  avgt      200    1462.327       33.817  ns/op
o.o.m.b.j.u.Base64Decode.testBase64WithErrorInputsDecode             144           4             32  avgt      200    1343.305       51.754  ns/op
o.o.m.b.j.u.Base64Decode.testBase64WithErrorInputsDecode             144           4             64  avgt      200    1572.225       37.234  ns/op
o.o.m.b.j.u.Base64Decode.testBase64WithErrorInputsDecode             144           4             80  avgt      200    1355.186       55.634  ns/op
o.o.m.b.j.u.Base64Decode.testBase64WithErrorInputsDecode             144           4             96  avgt      200    1351.123       55.995  ns/op
o.o.m.b.j.u.Base64Decode.testBase64WithErrorInputsDecode             144           4            112  avgt      200    1312.786       55.636  ns/op
o.o.m.b.j.u.Base64Decode.testBase64WithErrorInputsDecode             144           4            512  avgt      200    1415.921       56.890  ns/op
o.o.m.b.j.u.Base64Decode.testBase64WithErrorInputsDecode             144           4           1000  avgt      200    1509.619       64.826  ns/op
o.o.m.b.j.u.Base64Decode.testBase64WithErrorInputsDecode             144           4          20000  avgt      200    1367.791       53.885  ns/op
o.o.m.b.j.u.Base64Decode.testBase64WithErrorInputsDecode             144           4          50000  avgt      200    1399.333       71.046  ns/op
```

### CODE REVIEW

File: src/hotspot/cpu/x86/stubGenerator_x86_64.cpp [hunk 1]
```diff
@@ -5836,115 +5847,357 @@ address generate_avx_ghash_processBlocks() {
...
+    __ evmovdqaq(join01, ExternalAddress(StubRoutines::x86::base64_vbmi_join_0_1_addr()), Assembler::AVX_512bit, r13);
+    __ evmovdqaq(join12, ExternalAddress(StubRoutines::x86::base64_vbmi_join_1_2_addr()), Assembler::AVX_512bit, r13);
+    __ evmovdqaq(join23, ExternalAddress(StubRoutines::x86::base64_vbmi_join_2_3_addr()), Assembler::AVX_512bit, r13);
+
+    __ evmovdqaq(tmp16_op3, pack16_op, Assembler::AVX_512bit);
+    __ evmovdqaq(tmp16_op2, pack16_op, Assembler::AVX_512bit);
+    __ evmovdqaq(tmp16_op1, pack16_op, Assembler::AVX_512bit);
+    __ evmovdqaq(tmp32_op3, pack32_op, Assembler::AVX_512bit);
+    __ evmovdqaq(tmp32_op2, pack32_op, Assembler::AVX_512bit);
+    __ evmovdqaq(tmp32_op1, pack32_op, Assembler::AVX_512bit);
```
Comment (jatin-bhateja) [inline src/hotspot/cpu/x86/stubGenerator_x86_64.cpp]: Its a REG to REG move so aligned moves may be as good as unaligned moves.

File: src/hotspot/cpu/x86/stubGenerator_x86_64.cpp [hunk 2]
```diff
@@ -5836,115 +5847,357 @@ address generate_avx_ghash_processBlocks() {
...
 
     // load masks required for decoding data
     __ BIND(L_processdata);
-    __ vmovdqa64(join01, StubRoutines::x86::base64_vbmi_join_0_1_addr());
-    __ vmovdqa64(join12, StubRoutines::x86::base64_vbmi_join_1_2_addr());
-    __ vmovdqa64(join23, StubRoutines::x86::base64_vbmi_join_2_3_addr());
+    __ evmovdqaq(join01, ExternalAddress(StubRoutines::x86::base64_vbmi_join_0_1_addr()), Assembler::AVX_512bit, r13);
+    __ evmovdqaq(join12, ExternalAddress(StubRoutines::x86::base64_vbmi_join_1_2_addr()), Assembler::AVX_512bit, r13);
+    __ evmovdqaq(join23, ExternalAddress(StubRoutines::x86::base64_vbmi_join_2_3_addr()), Assembler::AVX_512bit, r13);
+
```
Comment (jatin-bhateja) [inline src/hotspot/cpu/x86/stubGenerator_x86_64.cpp]: Since you have already made sure to align the constants to be dumped at 64 byte aligned addresses, thus both aligned and unaligned moves should behave similar and same latency otherwise. Thus can't we use unaligned move instructions here given or did you see an exception even after inserting explicit align(64) before emitting constants.

File: src/hotspot/cpu/x86/stubGenerator_x86_64.cpp [hunk 3]
```diff
@@ -5836,115 +5847,357 @@ address generate_avx_ghash_processBlocks() {
...
+    __ evmovdquq(input0, Address(source, start_offset, Address::times_1, 0x00), Assembler::AVX_512bit);
+    __ evmovdquq(input1, Address(source, start_offset, Address::times_1, 0x40), Assembler::AVX_512bit);
+    __ evmovdquq(input2, Address(source, start_offset, Address::times_1, 0x80), Assembler::AVX_512bit);
+    __ evmovdquq(input3, Address(source, start_offset, Address::times_1, 0xc0), Assembler::AVX_512bit);
+
+    // Copy the low part of the lookup table into the destination of the permutation
+    __ evmovdqaq(translated0, lookup_lo, Assembler::AVX_512bit);
+    __ evmovdqaq(translated1, lookup_lo, Assembler::AVX_512bit);
+    __ evmovdqaq(translated2, lookup_lo, Assembler::AVX_512bit);
+    __ evmovdqaq(translated3, lookup_lo, Assembler::AVX_512bit);
```
Comment (jatin-bhateja) [inline src/hotspot/cpu/x86/stubGenerator_x86_64.cpp]: Same as above.

File: src/java.base/share/classes/java/util/Base64.java [hunk 4]
```diff
@@ -815,7 +815,7 @@ private int decode0(byte[] src, int sp, int sl, byte[] dst) {
                      * Calculate how many characters were processed by how many
                      * bytes of data were returned.
                      */
-                    int chars_decoded = (dl / 3) * 4;
+                    int chars_decoded = ((dl + 2) / 3) * 4;
```
Comment (jatin-bhateja) [inline src/java.base/share/classes/java/util/Base64.java:818]: 4 encoded bytes (24 actual bits + 8 bits pad) processed to  3 decoded bytes (24 bits), since the loop in decode block runs at the granularity of 4 bytes this changes is note clear to me.

File: src/hotspot/cpu/x86/assembler_x86.cpp [hunk 5]
```diff
@@ -3823,6 +3882,14 @@ void Assembler::evpermi2q(XMMRegister dst, XMMRegister nds, XMMRegister src, int
   emit_int16(0x76, (0xC0 | encode));
 }
 
+void Assembler::evpermt2b(XMMRegister dst, XMMRegister nds, XMMRegister src, int vector_len) {
+  assert(VM_Version::supports_evex(), "");
```
Comment (jatin-bhateja) [inline src/hotspot/cpu/x86/assembler_x86.cpp]: Check replacement VM_Version::supports_evex() -> VM_Version:supports_avx512_vbmi()

File: src/hotspot/cpu/x86/assembler_x86.cpp [hunk 6]
```diff
@@ -4091,6 +4158,15 @@ void Assembler::evpcmpeqq(KRegister kdst, XMMRegister nds, Address src, int vect
   emit_operand(as_Register(dst_enc), src);
 }
 
+void Assembler::evpmovb2m(KRegister kdst, XMMRegister src, int vector_len) {
+  assert(UseAVX > 2  && VM_Version::supports_avx512vlbw(), "");
```
Comment (jatin-bhateja) [inline src/hotspot/cpu/x86/assembler_x86.cpp]: UseAVX > 2 could be redundant here, its a JVM runtime flags, and this instruction is just a level level routine which should check existence of relevant hardware feature.  Also below assertion may not be needed.

File: src/hotspot/cpu/x86/assembler_x86.cpp [hunk 7]
```diff
@@ -4551,6 +4627,15 @@ void Assembler::vpmaddwd(XMMRegister dst, XMMRegister nds, XMMRegister src, int
   emit_int16((unsigned char)0xF5, (0xC0 | encode));
 }
 
+void Assembler::vpmaddubsw(XMMRegister dst, XMMRegister src1, XMMRegister src2, int vector_len) {
+  assert(vector_len == AVX_128bit ? VM_Version::supports_avx() :
+    (vector_len == AVX_256bit ? VM_Version::supports_avx2() :
+    (vector_len == AVX_512bit ? VM_Version::supports_evex() : 0)), "");
```
Comment (jatin-bhateja) [inline src/hotspot/cpu/x86/assembler_x86.cpp]: This check looks risky, if Vector Length is 128 bit we are simply bypassing  evex check, there is no guarantee that argument vector are not allocated from upper register bank ([XYZ]MM16-31) which mandates AVX512 support on the target. This is why for most of the instruction there are separate AVX512 variants. 

It should be ok to just add an evex version of the instruction for the time being.

File: src/hotspot/cpu/x86/assembler_x86.cpp [hunk 8]
```diff
@@ -4859,6 +4944,15 @@ void Assembler::vptest(XMMRegister dst, XMMRegister src, int vector_len) {
   emit_int16(0x17, (0xC0 | encode));
 }
 
+void Assembler::evptestmb(KRegister dst, XMMRegister nds, XMMRegister src, int vector_len) {
+  assert(VM_Version::supports_evex(), "");
```
Comment (jatin-bhateja) [inline src/hotspot/cpu/x86/assembler_x86.cpp]: Check replacement VM_Version::supports_evex() -> VM_Version:supports_avx512vlbw()

File: src/hotspot/cpu/x86/assembler_x86.cpp [hunk 9]
```diff
@@ -3008,6 +3008,65 @@ void Assembler::evmovdquq(Address dst, KRegister mask, XMMRegister src, bool mer
...
+  attributes.set_embedded_opmask_register_specifier(mask);
+  if (merge) {
+    attributes.reset_is_clear_context();
+  }
+  attributes.set_is_evex_instruction();
+  vex_prefix(dst, 0, src->encoding(), VEX_SIMD_66, VEX_OPCODE_0F, &attributes);
+  emit_int8(0x7F);
+  emit_operand(src, dst);
+}
+
```
Comment (jatin-bhateja) [inline src/hotspot/cpu/x86/assembler_x86.cpp]: In case we plan to remove aligned moves altogether as mentioned in a comment over stubGenerator_64.cpp then these can be removed.

File: src/hotspot/cpu/x86/stubGenerator_x86_64.cpp [hunk 10]
```diff
@@ -7205,6 +7458,15 @@ address generate_avx_ghash_processBlocks() {
...
       StubRoutines::_base64_encodeBlock = generate_base64_encodeBlock();
+      StubRoutines::x86::_lookup_lo = base64_vbmi_lookup_lo_addr();
+      StubRoutines::x86::_lookup_hi = base64_vbmi_lookup_hi_addr();
+      StubRoutines::x86::_lookup_lo_url = base64_vbmi_lookup_lo_url_addr();
+      StubRoutines::x86::_lookup_hi_url = base64_vbmi_lookup_hi_url_addr();
+      StubRoutines::x86::_pack_vec = base64_vbmi_pack_vec_addr();
+      StubRoutines::x86::_join_0_1 = base64_vbmi_join_0_1_addr();
+      StubRoutines::x86::_join_1_2 = base64_vbmi_join_1_2_addr();
+      StubRoutines::x86::_join_2_3 = base64_vbmi_join_2_3_addr();
+      StubRoutines::_base64_decodeBlock = generate_base64_decodeBlock();
```
Comment (jatin-bhateja) [inline src/hotspot/cpu/x86/stubGenerator_x86_64.cpp]: This block can be conditioned to existence of AVX512-VBMI since you are using some instructions needing that feature.

**Comment by jatin-bhateja** [comment]
Kindly add the pref results of  already existing JMH bench mark showcasing the gains. 
- test/micro/org/openjdk/bench/java/util/Base64Decode.java

File: src/java.base/share/classes/java/util/Base64.java (hunk 4 above)
Comment (asgibbons) [reply src/java.base/share/classes/java/util/Base64.java:818]: The length of the decoded string uses a ceil(dl / 3) operation.  The ((dl + 2) / 3) implements ceil(dl / 3).  Let's do an example:

Encode the 4 bytes "1234".  The result is a length 8 string of "MTIzNAo=".  Naturally, the length returned by the decoder will be 4.  The result of (4 / 3) * 4 will be 4, however the decode actually decoded 8 input bytes.  This causes the following instruction "sp += chars_decoded" to be incorrectly calculated, as the source offset should have been incremented by 8 and not 4.

Changing this to be ((dl + 2) / 3) * 4 yields ((4 + 2) / 3) * 4, which is 8 - the correct number of input bytes processed.

You can see a similar calculation in Base64Decode.java at line 80:

            byte[] dst = new byte[((srcLen + 2) / 3) * 4];

File: src/hotspot/cpu/x86/assembler_x86.cpp (hunk 9 above)
Comment (asgibbons) [reply src/hotspot/cpu/x86/assembler_x86.cpp]: OK.  I had thought that aligned moves from memory would be faster than unaligned, but I see this is only the case where the memory is not 64-byte aligned.  I'll remove these instructions and replace them with unaligned moves.

There IS a bug with the align() directive, however.  align(64) does *not* align on a 64-byte boundary.  I had to force it by specifying '    __ align(64, (unsigned long) __ pc());'.  Apparently align() uses the incorrect index for alignment.

File: src/hotspot/cpu/x86/assembler_x86.cpp (hunk 5 above)
Comment (asgibbons) [reply src/hotspot/cpu/x86/assembler_x86.cpp]: Done.

File: src/hotspot/cpu/x86/assembler_x86.cpp (hunk 6 above)
Comment (asgibbons) [reply src/hotspot/cpu/x86/assembler_x86.cpp]: Done.

File: src/hotspot/cpu/x86/assembler_x86.cpp (hunk 7 above)
Comment (asgibbons) [reply src/hotspot/cpu/x86/assembler_x86.cpp]: Changed to:

void Assembler::evpmaddubsw(XMMRegister dst, XMMRegister src1, XMMRegister src2, int vector_len) {
  assert(VM_Version::supports_avx512bw(), "");

File: src/hotspot/cpu/x86/assembler_x86.cpp (hunk 8 above)
Comment (asgibbons) [reply src/hotspot/cpu/x86/assembler_x86.cpp]: Done.

File: src/hotspot/cpu/x86/stubGenerator_x86_64.cpp (hunk 2 above)
Comment (asgibbons) [reply src/hotspot/cpu/x86/stubGenerator_x86_64.cpp]: Done.

File: src/hotspot/cpu/x86/stubGenerator_x86_64.cpp (hunk 1 above)
Comment (asgibbons) [reply src/hotspot/cpu/x86/stubGenerator_x86_64.cpp]: Done.

File: src/hotspot/cpu/x86/stubGenerator_x86_64.cpp (hunk 3 above)
Comment (asgibbons) [reply src/hotspot/cpu/x86/stubGenerator_x86_64.cpp]: Done.

File: src/hotspot/cpu/x86/stubGenerator_x86_64.cpp (hunk 10 above)
Comment (asgibbons) [reply src/hotspot/cpu/x86/stubGenerator_x86_64.cpp]: Done.

File: src/java.base/share/classes/java/util/Base64.java (hunk 4 above)
Comment (jatin-bhateja) [reply src/java.base/share/classes/java/util/Base64.java:818]: Input 4 byte = 32 bits this when encoded using base64 encoder i.e ceil(32 /6)×8 = 48 bits  I.e. 4bytes input encoded as 6 byte result.

**Comment by asgibbons** [comment]
TL;DR: My intrinsic properly handles padding bytes within an encoded string.

Let’s take the simplest case:  Encoding a single byte will produce 4 bytes of encoded data.  8 bits within the first two bytes decode to the single output byte, and the last two bytes are padding (‘==’).  So decoding these 4 bytes should return that a single byte was decoded to the destination.  I assume we can agree that this is the case.

Now the statement ‘int chars_decoded = (dl / 3) * 4;’  is calculating how many source (encoded) bytes were consumed in order to generate ‘dl’ bytes of decoded output.  If decodeBlock (correctly) returns that 1 byte of output was generated, this calculation states that the single byte was decoded using zero bytes of encoded data.  This is clearly incorrect, and leads to decode0 returning an incorrect decoded string length (in this case, it will return a value of 2, even though only a single byte was decoded).

This change effectively removes the restriction on the intrinsic of strictly returning an integer multiple of 3 bytes.  It has the added benefit of not affecting any currently-implemented intrinsic as well as properly following the RFC.

Thanks,
--Scott Gibbons
